server/src/main/java/org/opensearch/indices/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,241 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.indices;
+
+import org.roaringbitmap.RoaringBitmap;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    // This class shares a lot of the same fields with HybridIntKeyLookupStore, but basically none of the logic
+    // besides getters, so I decided against making it the superclass to HybridIntKeyLookupStore
+    protected final int modulo;
+    protected int size;
+    protected long memSizeCapInBytes;
+    protected int numAddAttempts;
+    protected int numCollisions;
+    protected boolean guaranteesNoFalseNegatives;
+    protected RoaringBitmap rbm;
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    protected final Lock readLock = lock.readLock();
+    protected final Lock writeLock = lock.writeLock();
+    protected RBMSizeEstimator sizeEstimator;
+    protected int maxNumEntries;
+    protected boolean atCapacity;
+
+    RBMIntKeyLookupStore(int modulo, long memSizeCapInBytes) {
+        this.modulo = modulo;
+        this.size = 0;
+        this.numAddAttempts = 0;
+        this.numCollisions = 0;
+        this.guaranteesNoFalseNegatives = true;
+        this.memSizeCapInBytes = memSizeCapInBytes; // A cap of 0 means no cap
+        this.sizeEstimator = new RBMSizeEstimator(modulo);
+        this.maxNumEntries = calculateMaxNumEntries();
+        this.rbm = new RoaringBitmap();
+    }
+
+    protected int calculateMaxNumEntries() {
+        if (memSizeCapInBytes == 0) {
+            return Integer.MAX_VALUE;
+        }
+        return sizeEstimator.getNumEntriesFromSizeInMB(RBMSizeEstimator.convertBytesToMB(memSizeCapInBytes));
+    }
+
+    protected final int transform(int value) {
+        return modulo == 0 ? value : value % modulo;
+    }
+
+    protected void handleCollisions(int transformedValue) {
+        numCollisions++;
+    }
+
+    @Override
+    public boolean add(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        writeLock.lock();
+        numAddAttempts++;
+        try {
+            if (size == maxNumEntries) {
+                atCapacity = true;
+                return false;
+            }
+            int transformedValue = transform(value);
+            boolean alreadyContained = contains(transformedValue);
```

Comment:
This seems wrong. We are passing transformedValue to contains, which internally again transforms it further. Shouldn't we be just passing contains(value) instead?

Reply:
Also considering this will mostly return false than true, as collisions would be less frequent, better to just call rbm.add() directly and avoid extra seek?

Reply:
You're right it's redundant, although transform() should be idempotent. I'll change that. But I think we need to keep the call to contains(), otherwise we won't know if there is a collision, and then we can't maintain our set of values which have had collisions, which we need to safely remove entries.